homeassistant/components/media_player/dlna_dmr.py

```diff
@@ -0,0 +1,806 @@
+    hass_data[name] = view
+
+
+def setup_platform(hass: HomeAssistant,
+                   config,
+                   add_devices,
+                   discovery_info=None):
+    """Set up DLNA DMR platform."""
+    if discovery_info and \
+       'upnp_device_type' in discovery_info and \
+       discovery_info['upnp_device_type'] != UPNP_DEVICE_MEDIA_RENDERER:
+        _LOGGER.debug('Device is not a MediaRenderer: %s',
+                      discovery_info.get('ssdp_description'))
+        return
+
+    if config.get(CONF_URL) is not None:
+        url = config.get(CONF_URL)
+        name = config.get(CONF_NAME)
+        udn = config.get(CONF_UDN)
+    elif discovery_info is not None:
+        url = discovery_info['ssdp_description']
+        name = discovery_info['name']
+        udn = discovery_info['udn']
+
+    # set up our Views, if not already done so
+    if __name__ not in hass.data:
+        hass.data[__name__] = {}
+
+    # ensure view has been started
+    hass.async_run_job(start_notify_view, hass)
+    while 'notify_view' not in hass.data[__name__]:
+        time.sleep(0.1)
+    notify_view = hass.data[__name__]['notify_view']
+
+    # create device
+    from async_upnp_client import UpnpFactory
+    requester = HassUpnpRequester(hass)
+    factory = UpnpFactory(requester, disable_state_variable_validation=True)
+    device = DlnaDmrDevice(hass, url, udn, name, factory, notify_view)
+
+    _LOGGER.debug("Adding device: %s", device)
+    add_devices([device])
+
+
+async def fetch_headers(hass, url, headers):
+    """Fetch headers from URL, first by trying HEAD, then by trying a GET."""
+    # try a HEAD request to the source
+    src_response = None
+    try:
+        session = async_get_clientsession(hass)
+        src_response = await session.head(url, headers=headers)
+        await src_response.release()
+    except aiohttp.ClientError:
+        pass
+
+    if src_response and 200 <= src_response.status < 300:
+        return src_response.headers
+
+    # try a GET request to the source, but ignore all the data
+    session = async_get_clientsession(hass)
+    src_response = await session.get(url, headers=headers)
+    await src_response.release()
+
+    return src_response.headers
+
+
+class UpnpNotifyView(HomeAssistantView):
+    """Callback view for UPnP NOTIFY messages."""
+
+    url = '/api/dlna_dmr.notify'
+    name = 'api:dlna_dmr:notify'
+    requires_auth = False
+
+    def __init__(self, hass):
+        """Initializer."""
+        self.hass = hass
+        self._registered_services = {}
+        self._backlog = {}
+
+    def register(self, router):
+        """Register the view with a router."""
+        handler = request_handler_factory(self, self.async_notify)
+        router.add_route('notify', UpnpNotifyView.url, handler)
+
+    async def async_notify(self, request):
+        """Callback method for NOTIFY requests."""
+        if 'SID' not in request.headers:
+            return aiohttp.web.Response(status=412)
```

Comment:
I think it'd be nice to define non-standard (or more like less commonly used..) response codes as constants. Personally I couldn't say what are 202 nor 412 without looking those up.

Reply:
The 412 comes from the UPnP spec. I'll turn them into constants and use those.

Reply:
I see some constants in homeassistant/const.py, I'll add the ones I use there and use those.

Reply:
Better add these to this component only (or the upstream lib, and import them), as they are probably not very interesting for other homeassistant components.